pkg/olog/olog.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Implements the public API for the olog package.
+
+// Package olog implements a lightweight logging library built around
+// the slog package. It aims to never mask the core slog.Logger type by
+// default. Provided is a global system for controlling logging levels
+// based on the package and module that a logger was created in, with a
+// system to update the logging level at runtime.
+//
+// This package does not provide the ability to ship logs to a remote
+// server, instead a logging collector should be used.
+package olog
+
+import (
+	"fmt"
+	"log/slog"
+
+	"github.com/getoutreach/gobox/pkg/callerinfo"
+)
+
+// New creates a new slog instance that can be used for logging. The
+// provided logger use the global handler provided by this package. See
+// the documentation on the 'handler' global for more information.
+//
+// The logger will be automatically associated with the module and
+// package that it was instantiated in. This is done by looking at the
+// call stack.
+//
+// Note: As mentioned above, this logger is associated with the module
+// and package that created it. So, if you pass this logger to another
+// module or package, the association will NOT be changed. This
+// includes the caller metadata added to every log line as well as
+// log-level management. If a type has a common logging format that the
+// other module or package should use, then a slog.LogValuer should be
+// implemented on that type instead of passing a logger around. If
+// trying to set attributes the be logged by default, this is not
+// supported without retaining the original association.
+func New() *slog.Logger {
+	m, err := getMetadata()
+	if err != nil {
+		//nolint:errorlint // Why: We can't wrap panic-d errors.
+		panic(fmt.Errorf("failed to get information about what created the logger: %v", err))
```

Comment:
I really don't like the idea of such a standard and foundational library panicking like this. Why can't we just return an error from New or default the metadata in the event of an error and just log the error through the `fmt` stdlib package?

Reply:
Talked offline, added a comment denoting why we `panic` here.